### PR TITLE
dm: add dml replication logic

### PR DESCRIPTION
**Pull request by ran-huang**
### What is changed, added or deleted? (Required)

Add a doc for DML replication logic.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v6.1 (TiDB 6.1 versions)
- [x] v6.0 (TiDB 6.0 versions)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: dm/dm-dml-replication-logic.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+DML 执行和 checkpoint 写操作不是同步的，并且写 checkpoint 操作和写下游数据也并不能保证原子性。当 DM 因为某些原因异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点。因此，当同步任务重启时，DM 可能会重复写入部分数据，也就是说，DM 实际上提供的是“至少一次处理”的逻辑（At-least-once processing），相同的数据可能会被处理一次以上。
+
+为了保证数据是可重入的，DM 在异常重启时会进入安全模式。具体逻辑参阅 [DM 安全模式](/dm/dm-safe-mode.md)。
+
+开启安全模式期间，为了保证数据可重入，DM 会进行如下转换：
+
+* 将上游 `INSERT` 语句，转换成 `REPLACE` 语句
+* 将上游 `UPDATE` 语句，转换成 `DELETE` + `REPLACE` 语句。
+
+### 精确一次处理 (Exactly-Once Processing)
```
Comment (ran-huang) [inline dm/dm-dml-replication-logic.md:135]: 这段建议删掉

File: TOC.md [hunk 2]
```diff
@@ -351,6 +351,8 @@
           - [DM 架构简介](/dm/dm-arch.md)
           - [DM-worker 说明](/dm/dm-worker-intro.md)
           - [Relay Log](/dm/relay-log.md)
+        - 同步机制
```
Comment (sunzhaoyang) [inline TOC.md:354]: 这个二级菜单我觉得要再讨论下。一方面有些文档我直接放在架构下面，不怎么合适。另一方面”同步机制“貌似只会有这么一两篇文档，也比较鸡肋。

或者我们考虑改成 内部机制，运行机制，或者其他的名字，把其他分类不太好的也移过来

File: dm/dm-dml-replication-logic.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
+---
+title: Data Migration 中的 DML 同步机制
+summary: 本文介绍了 DM 核心处理单元 Sync 如何同步 DML 语句。
+---
+
+# Data Migration 中的 DML 同步机制
+
+本文介绍了 DM 核心处理单元 Sync 如何同步从数据源或 relay log 中读取到的 DML。本文将介绍 Binlog 的整体处理流程，包括 binlog 读取、过滤、路由、转换、优化以及执行等逻辑，并详细解读 DML 优化逻辑和 DML 执行逻辑。
```
Comment (sunzhaoyang) [inline dm/dm-dml-replication-logic.md:8]:
```suggestion
本文介绍了 DM 核心处理单元 Sync 如何同步从数据源或 relay log 中读取到的 DML。本文将介绍 DML 事件在 DM 内部的完整处理流程，包括 binlog 读取、过滤、路由、转换、优化以及执行等逻辑，并详细解读 DML 优化逻辑和 DML 执行逻辑。
```

File: dm/dm-dml-replication-logic.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+在 DM 中，checkpoint 默认每 30 秒更新一次。同时，由于存在多个 DML worker 进程，checkpoint 进程会计算所有 DML worker 中同步进度最晚的 binlog 位点，将该位点作为当前同步的 checkpoint。所有早于此位点的 binlog，都已保证被成功地执行到下游。
+
+关于 Checkpoint 的详细信息，参考 [Checkpoint](/dm/dm-checkpoint.md)。
+
+## 注意事项
+
+### 事务一致性
+
+DM 是按照“行级别”进行数据同步的。在 DM 中，上游的一个事务会被拆成多行，分发到不同的 DML Worker 中并发执行。因此，当 DM 同步任务报错暂停，或者用户手动暂停任务时，下游可能停留在一个中间状态；即上游一个事务中的 DML 语句，可能一部分同步到下游，一部分没有，导致下游处于不一致的状态。
```
Comment (sunzhaoyang) [inline dm/dm-dml-replication-logic.md:138]:
```suggestion
DM 是按照“行级别”进行数据同步的，并不保证事务原样同步。在 DM 中，上游的一个事务会被拆成多行，分发到不同的 DML Worker 中并发执行。因此，当 DM 同步任务报错暂停，或者用户手动暂停任务时，下游可能停留在一个中间状态；即上游一个事务中的 DML 语句，可能一部分同步到下游，一部分没有，导致下游处于不一致的状态。
```

File: dm/dm-dml-replication-logic.md (hunk 1 above)
Comment (sunzhaoyang) [reply dm/dm-dml-replication-logic.md:153]: 我觉得不用删。内容可以改为尚未支持。这样可以减少一些猜测和询问。 @GMHDBJD

File: dm/dm-dml-replication-logic.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+DM 将多条 DML 攒到一个事务中执行到下游。当 DML Worker 收到 DML 时，将 DML 加入到缓存中。当缓存中 DML 数量达到预定阈值时，或者 DML worker 较长时间没有收到 DML 时，DML worker 将缓存中的 DML 执行到下游。
+
+你可以通过 [`syncer.batch` 配置项](/dm/dm-tune-configuration.md#batch)，修改每个事务包含的 DML 的数量。
+
+### checkpoint
+
+DML 执行和 checkpoint 更新的操作不是原子的。
+
+在 DM 中，checkpoint 默认每 30 秒更新一次。同时，由于存在多个 DML worker 进程，checkpoint 进程会计算所有 DML worker 中同步进度最晚的 binlog 位点，将该位点作为当前同步的 checkpoint。所有早于此位点的 binlog，都已保证被成功地执行到下游。
```
Comment (lance6716) [inline dm/dm-dml-replication-logic.md:130]:
```suggestion
在 DM 中，checkpoint 默认每 30 秒更新一次。同时，由于存在多个 DML worker 进程，checkpoint 进程会计算所有 DML worker 中同步进度最早的 binlog 位点，将该位点作为当前同步的 checkpoint。所有早于此位点的 binlog，都已保证被成功地执行到下游。
```

File: dm/dm-dml-replication-logic.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+关于 Checkpoint 的详细信息，参考 [Checkpoint](/dm/dm-checkpoint.md)。
+
+## 注意事项
+
+### 事务一致性
+
+DM 是按照“行级别”进行数据同步的。在 DM 中，上游的一个事务会被拆成多行，分发到不同的 DML Worker 中并发执行。因此，当 DM 同步任务报错暂停，或者用户手动暂停任务时，下游可能停留在一个中间状态；即上游一个事务中的 DML 语句，可能一部分同步到下游，一部分没有，导致下游处于不一致的状态。
+
+为了尽可能使任务暂停时下游处于一致状态，DM v5.3.0 起，在任务暂停时会等待上游事务全部同步到下游后，才真正暂停任务。这个等待时间为 10 秒，如果上游一个事务在 10 秒内还未全部同步到下游，那么下游仍然可能处于不一致的状态。
```
Comment (lance6716) [inline dm/dm-dml-replication-logic.md:140]:
```suggestion
为了尽可能使任务暂停时下游处于一致状态，DM v5.3.0 起，在任务暂停时会等待正在处理的上游事务全部同步到下游后，才真正暂停任务。这个等待时间为 10 秒，如果上游一个事务在 10 秒内还未全部同步到下游，那么下游仍然可能处于不一致的状态。
```

File: dm/dm-dml-replication-logic.md (hunk 1 above)
Comment (lance6716) [reply dm/dm-dml-replication-logic.md:153]: 我看完这一段，没发现明确说出来 dm 不支持精确一次同步就结束了。还以为是这一段没写完呢

File: TOC.md (hunk 2 above)
Comment (ran-huang) [reply TOC.md:354]: “内部机制”或者“原理”？

Comment (sunzhaoyang) [reply TOC.md:354]: "运行机制" or ”原理解读“? 你喜欢哪个？

File: dm/dm-dml-replication-logic.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+为了保证数据是可重入的，DM 在异常重启时会进入安全模式。具体逻辑参阅 [DM 安全模式](/dm/dm-safe-mode.md)。
+
+开启安全模式期间，为了保证数据可重入，DM 会进行如下转换：
+
+* 将上游 `INSERT` 语句，转换成 `REPLACE` 语句
+* 将上游 `UPDATE` 语句，转换成 `DELETE` + `REPLACE` 语句。
+
+### 精确一次处理 (Exactly-Once Processing)
+
+DM 的拆分事务再并发同步的逻辑存在已知问题，比如下游可能停在一个不一致的状态，数据的同步顺序与上游不一致，可能导致数据重入。安全模式期间 `REPLACE` 语句会有一定的性能损失，如果下游需要捕获数据变更（如 change data capture），那么重复处理也不可接受。
```
Comment (sunzhaoyang) [inline dm/dm-dml-replication-logic.md:155]:
```suggestion
目前 DM 仅保证最终一致性，尚未支持”精准一次处理“及”保持事务原有顺序同步“。
```

Comment (sunzhaoyang) [reply dm/dm-dml-replication-logic.md:155]: @lance6716 @GMHDBJD

Comment (GMHDBJD) [reply dm/dm-dml-replication-logic.md:155]: lgtm

File: TOC.md (hunk 2 above)
Comment (ran-huang) [reply TOC.md:354]: 运行机制吧

**Review by lance6716 — APPROVED** [review]

File: TOC.md (hunk 2 above)
Comment (sunzhaoyang) [inline TOC.md:354]:
```suggestion
        - 运行机制
```

File: dm/dm-dml-replication-logic.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+你可以通过 [`syncer.batch` 配置项](/dm/dm-tune-configuration.md#batch)，修改每个事务包含的 DML 的数量。
+
+### checkpoint
+
+DML 执行和 checkpoint 更新的操作不是原子的。
+
+在 DM 中，checkpoint 默认每 30 秒更新一次。同时，由于存在多个 DML worker 进程，checkpoint 进程会计算所有 DML worker 中同步进度最早的 binlog 位点，将该位点作为当前同步的 checkpoint。所有早于此位点的 binlog，都已保证被成功地执行到下游。
+
+关于 Checkpoint 的详细信息，参考 [Checkpoint](/dm/dm-checkpoint.md)。
```
Comment (ran-huang) [inline dm/dm-dml-replication-logic.md:132]:
```suggestion
<!-- 关于 Checkpoint 的详细信息，参考 Checkpoint /dm/dm-checkpoint.md。-->
```
dm-checkpoint.md 还没合并，暂时隐藏掉。

File: dm/dm-dml-replication-logic.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
+---
+title: Data Migration 中的 DML 同步机制
+summary: 本文介绍了 DM 核心处理单元 Sync 如何同步 DML 语句。
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:3]:
```suggestion
summary: 了解 DM 核心处理单元 Sync 如何同步 DML 语句。
```

File: dm/dm-dml-replication-logic.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+DELETE + INSERT => UPDATE
+```
+
+Compactor 特性默认关闭，如需启用可在迁移任务的`sync`配置模块开启，例如：
+
+```
+syncers:                             # sync 处理单元的运行配置参数
+  global:                            # 配置名称
+    ...                              # 省略其他配置
+    compact: false
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:54]: 我理解开启某个特性，应该是将值设为 true？
```suggestion
    compact: false
```

File: dm/dm-dml-replication-logic.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+```
+INSERT + UPDATE => INSERT
+INSERT + DELETE => DELETE
+UPDATE + UPDATE => UPDATE
+UPDATE + DELETE => DELETE
+DELETE + INSERT => UPDATE
+```
+
+Compactor 特性默认关闭，如需启用可在迁移任务的`sync`配置模块开启，例如：
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:48]:
```suggestion
Compactor 特性默认关闭，如需启用可在迁移任务的 `sync` 配置模块开启，例如：
```

File: dm/dm-dml-replication-logic.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  UPDATE tb SET a=1, b=1 WHERE a=1;
++ UPDATE tb SET a=2, b=2 WHERE a=2;
+= INSERT tb(a,b) VALUES(1,1),(2,2) ON DUPLICATE UPDATE a=VALUES(a), b=VALUES(b)
+
+  DELETE tb WHERE a=1
++ DELETE tb WHERE a=2
+= DELETE tb WHERE (a) IN (1),(2);
+```
+
+Merger 特性默认关闭，如需启用可在迁移任务的`sync`配置模块开启，例如：
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:83]:
```suggestion
Merger 特性默认关闭，如需启用可在迁移任务的 `sync` 配置模块开启，例如：
```

File: dm/dm-dml-replication-logic.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+### Worker count
+
+Causality 可以通过冲突检测算法将 binlog 分成多个 group 并发地执行到下游。DM 通过设置 worker-count，控制并发的数量。当下游 TiDB 的 CPU 占用不高时，增大并发的数量可以有效地提高数据同步的吞吐量。
+
+你可以通过 [`syncer.worker-count` 配置项](/dm/dm-tune-configuration.md#worker-count)，修改并发迁移 DML 的线程数量。
+
+### Batch
+
+DM 将多条 DML 攒到一个事务中执行到下游。当 DML Worker 收到 DML 时，将 DML 加入到缓存中。当缓存中 DML 数量达到预定阈值时，或者 DML worker 较长时间没有收到 DML 时，DML worker 将缓存中的 DML 执行到下游。
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:122]:
```suggestion
DM 将多条 DML 积累到一个事务中执行到下游。当 DML Worker 收到 DML 时，将 DML 加入到缓存中。当缓存中 DML 数量达到预定阈值时，或者 DML worker 较长时间没有收到 DML 时，DML worker 将缓存中的 DML 执行到下游。
```

File: dm/dm-dml-replication-logic.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+<!-- 关于 Checkpoint 的详细信息，参考 Checkpoint /dm/dm-checkpoint.md。-->
+
+## 注意事项
+
+### 事务一致性
+
+DM 是按照“行级别”进行数据同步的，并不保证事务原样同步。在 DM 中，上游的一个事务会被拆成多行，分发到不同的 DML Worker 中并发执行。因此，当 DM 同步任务报错暂停，或者用户手动暂停任务时，下游可能停留在一个中间状态；即上游一个事务中的 DML 语句，可能一部分同步到下游，一部分没有，导致下游处于不一致的状态。
+
+为了尽可能使任务暂停时下游处于一致状态，DM v5.3.0 起，在任务暂停时会等待正在处理的上游事务全部同步到下游后，才真正暂停任务。这个等待时间为 10 秒，如果上游一个事务在 10 秒内还未全部同步到下游，那么下游仍然可能处于不一致的状态。
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:140]:
```suggestion
为了尽可能使任务暂停时下游处于一致状态，DM v5.3.0 起，任务暂停时，会等待 10 秒，使正在处理的上游事务全部同步到下游再真正暂停任务。如果上游一个事务在 10 秒内还未全部同步到下游，那么下游仍然可能处于不一致的状态。
```

File: dm/dm-dml-replication-logic.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+### 安全模式
+
+DML 执行和 checkpoint 写操作不是同步的，并且写 checkpoint 操作和写下游数据也并不能保证原子性。当 DM 因为某些原因异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点。因此，当同步任务重启时，DM 可能会重复写入部分数据，也就是说，DM 实际上提供的是“至少一次处理”的逻辑（At-least-once processing），相同的数据可能会被处理一次以上。
+
+为了保证数据是可重入的，DM 在异常重启时会进入安全模式。具体逻辑参阅 [DM 安全模式](/dm/dm-safe-mode.md)。
+
+开启安全模式期间，为了保证数据可重入，DM 会进行如下转换：
+
+* 将上游 `INSERT` 语句，转换成 `REPLACE` 语句
+* 将上游 `UPDATE` 语句，转换成 `DELETE` + `REPLACE` 语句。
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:151]:
```suggestion
* 将上游 `INSERT` 语句转换成 `REPLACE` 语句
* 将上游 `UPDATE` 语句转换成 `DELETE` + `REPLACE` 语句。
```

File: dm/dm-dml-replication-logic.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+### 精确一次处理 (Exactly-Once Processing)
+
+目前 DM 仅保证最终一致性，尚未支持”精确一次处理“及”保持事务原有顺序同步“。
```
Comment (shichun-0415) [inline dm/dm-dml-replication-logic.md:155]:
```suggestion
目前 DM 仅保证最终一致性，尚未支持“精确一次处理”及“保持事务原有顺序同步”。
```

**Review by shichun-0415 — APPROVED** [review]
@ran-huang I have completed review and left some minor modificiations. PTAL.

File: dm/dm-dml-replication-logic.md (hunk 10 above)
Comment (sunzhaoyang) [reply dm/dm-dml-replication-logic.md:54]: 配置里写的都是默认值

File: dm/dm-dml-replication-logic.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+DELETE + INSERT => UPDATE
+```
+
+Compactor 特性默认关闭，如需启用可在迁移任务的 `sync` 配置模块开启，例如：
+
+```
+syncers:                             # sync 处理单元的运行配置参数
+  global:                            # 配置名称
+    ...                              # 省略其他配置
+    compact: false
```
Comment (sunzhaoyang) [inline dm/dm-dml-replication-logic.md:54]:
```suggestion
    compact: true
```

**Review by sunzhaoyang — APPROVED** [review]
LGTM

**Pull request merged**